validator/src/main/java/com/google/daq/mqtt/sequencer/SequenceBase.java

```diff
@@ -1252,47 +1255,33 @@ protected void updateConfig(String reason) {
 
   private void updateConfig(String reason, boolean force) {
     updateConfig(reason, force, true);
-
   }
 
   private void updateConfig(String reason, boolean force, boolean waitForSync) {
     assertConfigIsNotPending();
 
-    // Add a forced sleep to make sure second-quantized timestamps are unique.
-    safeSleep(CONFIG_BARRIER_MS);
-
     if (doPartialUpdates && !force) {
-      updateConfig(SubFolder.SYSTEM, augmentConfig(deviceConfig.system));
-      updateConfig(SubFolder.POINTSET, deviceConfig.pointset);
-      updateConfig(SubFolder.GATEWAY, deviceConfig.gateway);
-      updateConfig(SubFolder.LOCALNET, deviceConfig.localnet);
-      updateConfig(SubFolder.BLOBSET, deviceConfig.blobset);
-      updateConfig(SubFolder.DISCOVERY, deviceConfig.discovery);
+      updateConfig(reason, waitForSync, SubFolder.SYSTEM, augmentConfig(deviceConfig.system));
+      updateConfig(reason, waitForSync, SubFolder.POINTSET, deviceConfig.pointset);
+      updateConfig(reason, waitForSync, SubFolder.GATEWAY, deviceConfig.gateway);
+      updateConfig(reason, waitForSync, SubFolder.LOCALNET, deviceConfig.localnet);
+      updateConfig(reason, waitForSync, SubFolder.BLOBSET, deviceConfig.blobset);
+      updateConfig(reason, waitForSync, SubFolder.DISCOVERY, deviceConfig.discovery);
     } else {
       if (force) {
         debug("Forcing config update");
         sentConfig.remove(UPDATE);
       }
-      updateConfig(UPDATE, deviceConfig);
-    }
-
-    if (!waitForSync) {
-      waitForConfigIsNotPending();
-    } else if (configIsPending()) {
-      configStateStart = catchToNull(() -> deviceState.timestamp);
-      debug(format("Saving pre-config state timestamp " + isoConvert(configStateStart)));
-      lastConfigUpdate = CleanDateFormat.clean(Instant.now());
-      String debugReason = reason == null ? "" : (", because " + reason);
-      debug(format("Update lastConfigUpdate %s%s", lastConfigUpdate, debugReason));
-      waitForConfigSync();
+      updateConfig(reason, waitForSync, UPDATE, deviceConfig);
     }
 
     assertConfigIsNotPending();
 
     captureConfigChange(reason);
   }
 
-  private boolean updateConfig(SubFolder subBlock, Object data) {
+  private boolean updateConfig(String reason, boolean waitForSync, SubFolder subBlock,
```

Comment:
Okay, I kind of get this piece but would still like to share my understanding - We are now waiting for at least CONFIG_BARRIER seconds before making another config update on the device. 

Sequencer needs to validate each config update and that is why we care about not losing any?

When applying config to any device generally, we would not care about losing any intermediate ones, right? Because only the latest one sticks?

Reply:
I made the config barrer logic dependent on the log-level setting (easiest way for now). So if it's TRACE then it will gate the config, but if it's not then it just treats it as per normal.  In the future we can make that more configurable somehow if needed.